### PR TITLE
sql: resolve function names as all other objects

### PR DESCRIPTION
**Pull request by sploiselle**
Ultimately, we need to be able to track functions as objects that users depend on. This is necessary for:
- [UDFs](https://github.com/MaterializeInc/materialize/pull/7346#discussion_r668954450)
- Preventing users from creating objects that depend on unstable catalog objects without having a flag set
 
To be able to track functions as dependencies, we can resolve functions as all other objects (this is currently where we track all dependencies).

Other approaches considered included determining these dependencies _after_ planning, however that meant that we needed to return the `GlobalId`s in the plans and that seemed crazy.

One cost to this change is that we lose "nice error messages" when trying to find functions that do not exist--the new error messages in this case will not include the arguments' types. I'd argue that this is a net improvement because it more clearly delineates between a "catalog error" and a "function resolution" error, but it does diverge from PG's error strategy.

One last note: this PR is fairly ancient and from a different era when design docs were less prevalent, so I never wrote a design doc for this. However, I was able to get this green with a very narrowly defined change in behavior so hope I can get away without writing one.

### Motivation

This PR adds a known-desirable feature. https://github.com/MaterializeInc/materialize/pull/7346#discussion_r668954450

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by sploiselle** [comment]
@benesch I have this in a place where the failures are largely cosmetic/intended. One unclear bit is handling e.g. calls to `mz_internal.mz_sleep`. For instance, we use this code:

```
> SET cluster=cluster1

> CREATE TABLE sleep_table (sleep INTEGER);

> CREATE MATERIALIZED VIEW sleep_view AS SELECT mz_internal.mz_sleep(sleep) FROM sleep_table;

> INSERT INTO sleep_table SELECT 1200 FROM generate_series(1,32)
```

to incapacitate `cluster1`--I'm not sure how else we can facilitate that kind of behavior without `mz_sleep`. Do we want to:
- move `mz_sleep` into `mz_catalog`
- set an option on `CREATE MATERIALIZED VIEW` to allow `mz_internal` dependencies
- something else?

**Comment by benesch** [comment]
> allow unstable dependencies in unsafe mode

This sounds like the best option to me!

**Comment by sploiselle** [comment]
@jkosh44 someone from surfaces should review; @mjibson somewhat volunteered

**Comment by jkosh44** [comment]
> @jkosh44 someone from surfaces should review; @mjibson somewhat volunteered

In that case I'll leave it to @mjibson

File: src/sql/src/plan/transform_ast.rs
Comment (mjibson) [inline src/sql/src/plan/transform_ast.rs:301]: Should use `PG_CATALOG_SCHEMA` in all places instead of hard coding a string.

File: src/sql/src/plan/statement.rs
Comment (mjibson) [inline src/sql/src/plan/statement.rs:896]: Can this get some words about what is dangerous about it so it's clear even to us when it's appropriate to use?

File: src/sql/src/plan/query.rs
Comment (mjibson) [inline src/sql/src/plan/query.rs:4542]: Should this whole block just go away?

File: src/sql/src/plan/query.rs
Comment (mjibson) [inline src/sql/src/plan/query.rs]: Remove?

File: src/sql/src/normalize.rs
Comment (mjibson) [inline src/sql/src/normalize.rs:183]: Remove?

File: src/sql/src/func.rs
Comment (mjibson) [inline src/sql/src/func.rs]: Why was this added?

File: src/sql/src/query_model/test/catalog.rs
Comment (mjibson) [inline src/sql/src/query_model/test/catalog.rs:188]: We did some work to fully remove the DummyCatalog. Is it worthwhile to add another test impl back? Is the current `Catalog::open_debug_*` stuff still to heavy because it requires a running cockroach?

File: src/sql/src/plan/query.rs
Comment (mjibson) [inline src/sql/src/plan/query.rs:4219]: Why do we know `name` will always resolve? Is that guaranteed by some caller?

File: src/sql/src/func.rs
Comment (mjibson) [inline src/sql/src/func.rs:968]: There's a lot of extra expects on humanize_resolved_name now, and I'm a bit scared of those. Why are they safe?

**Review by mjibson — APPROVED** [review]
Looks great! No major questions, and if the tests pass let's get this in.

**Review by mjibson — APPROVED** [review]
Looks great! No major questions, and if the tests pass let's get this in.

**Comment by def-** [comment]
Would it be possible to rebase this change? Want to take a look from QA side, there are some conflicts though.

**Comment by sploiselle** [comment]
@def- I likely will not work on rebasing this until Friday.

File: src/sql-parser/src/parser.rs
Comment (sploiselle) [reply src/sql-parser/src/parser.rs:398]: This allows `SELECT [id AS foo]()`; there are parser tests for it, as well. https://github.com/MaterializeInc/materialize/pull/15719/files#diff-a26d784f9eb37b7f351491aafeac2d3ac77681832131c139b33bf115acb44b24R24

File: src/sql/src/plan/query.rs
Comment (sploiselle) [reply src/sql/src/plan/query.rs:4219]: This is guaranteed by the new name resolution pass over functions.

File: src/sql/src/func.rs
Comment (sploiselle) [reply src/sql/src/func.rs:968]: The name resolution pass now guarantees that any function name we receive exists in the catalog.

File: src/sql/src/plan/transform_ast.rs
Comment (sploiselle) [reply src/sql/src/plan/transform_ast.rs]: Did this and a little bit more by moving these consts into `mz_repr::namespaces`

File: src/sql/src/plan/statement.rs
Comment (sploiselle) [reply src/sql/src/plan/statement.rs]: Done

File: src/sql/src/plan/query.rs
Comment (sploiselle) [reply src/sql/src/plan/query.rs]: Derp, yes

File: src/sql/src/plan/query.rs
Comment (sploiselle) [reply src/sql/src/plan/query.rs]: Sorry, yep

File: src/sql/src/normalize.rs
Comment (sploiselle) [reply src/sql/src/normalize.rs]: So many derps!

File: src/sql/src/query_model/test/catalog.rs
Comment (sploiselle) [reply src/sql/src/query_model/test/catalog.rs]: This was just merge skew; removed

File: src/sql/src/names.rs
Comment (sploiselle) [reply src/sql/src/names.rs:1116]: Yes! Deleted elsewhere

**Comment by sploiselle** [comment]
@def- Rebased this; I am eager to get this merged given the likelihood of more merge conflicts. This is a pretty foundational part of Materialize, so it being green is a pretty large stress test. I'll leave this open until Monday, and merge then; that will give us the week for any additional testing + runs of nightlies.

**Comment by sploiselle** [comment]
@MaterializeInc/qa Actually, could one of you all kick off a nightly run on this so there are fewer surprises after merging?

**Comment by def-** [comment]
Sure, kicked off coverage and nightly:
https://buildkite.com/materialize/coverage/builds/108
https://buildkite.com/materialize/nightlies/builds/2362

**Review by def- — CHANGES_REQUESTED** [review]
[SQLsmith](https://buildkite.com/materialize/nightlies/builds/2362#018811fa-8762-42f3-8899-8d8f3cd0932d) is failing a lot:
```
--- [SQLsmith] Error XX000: ERROR:  function array_remove(mz_aclitem[], text) does not exist
```
The query is using it via `pg_catalog.array_remove(...)`. Is this only expected for `list_*` and `array_*` functions? According to `select * from mz_functions` the functions should still be in `pg_catalog`, so my expectation was that this should keep working.

File: src/adapter/src/catalog/builtin_table_updates.rs
Comment (def-) [inline src/adapter/src/catalog/builtin_table_updates.rs:1277]: Is this block covered by a test? Looks interesting to test

File: src/adapter/src/catalog/builtin_table_updates.rs
Comment (def-) [inline src/adapter/src/catalog/builtin_table_updates.rs:1285]: Is this block covered by a test? Looks interesting to test

File: src/sql/src/plan/query.rs
Comment (def-) [inline src/sql/src/plan/query.rs:2097]: This appears to be only tested in unit tests, might consider if it's testable in end-to-end test too.

**Review by def- — COMMENTED** [review]
I put the interesting uncovered blocks from coverage report as inline comments based on https://buildkite.com/materialize/coverage/builds/108#_

I hope it's helpful.

File: src/sql/src/plan/query.rs
Comment (sploiselle) [reply src/sql/src/plan/query.rs:2097]: The change this PR introduces is how the table's identified; the original change for this work included extensive tests https://github.com/MaterializeInc/materialize/pull/9988

File: src/adapter/src/catalog/builtin_table_updates.rs
Comment (sploiselle) [reply src/adapter/src/catalog/builtin_table_updates.rs]: This was re-added by merge skew.

**Review by def- — APPROVED** [review]
Looks good to me, thanks for adjusting the SQLsmith known errors

**Pull request merged**
